### PR TITLE
config: metadata life time with Any conversion

### PR DESCRIPTION
**Pull request by botengyao**
One fromResource variant is taking the `ProtobufWkt::Any& resource` as the input, and then a temp value is introduced in this static method.
```
      envoy::service::discovery::v3::Resource r;
      MessageUtil::unpackTo(resource, r); 
```
A ref to a temp val can cause `segv` fault when the temp is distroied.

Commit Message: 
Additional Description:
Risk Level: Low
Testing: unit test
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by botengyao** [comment]
cc @jmarantz

File: source/common/config/decoded_resource_impl.h [hunk 1]
```diff
@@ -101,7 +101,7 @@ class DecodedResourceImpl : public DecodedResource {
 
   // This is the metadata info under the Resource wrapper.
   // It is intended to be consumed in the xds_config_tracker extension.
-  const OptRef<const envoy::config::core::v3::Metadata> metadata_;
+  const absl::optional<envoy::config::core::v3::Metadata> metadata_;
```
Comment (jmarantz) [inline source/common/config/decoded_resource_impl.h:104]: I think this change is needed.

File: envoy/config/subscription.h [hunk 2]
```diff
@@ -62,9 +62,9 @@ class DecodedResource {
   virtual bool hasResource() const PURE;
 
   /**
-   * @return optional ref<envoy::config::core::v3::Metadata> of a resource.
+   * @return optional envoy::config::core::v3::Metadata of a resource.
    */
-  virtual const OptRef<const envoy::config::core::v3::Metadata> metadata() const PURE;
+  virtual const absl::optional<envoy::config::core::v3::Metadata> metadata() const PURE;
```
Comment (jmarantz) [inline envoy/config/subscription.h:67]: I am not understanding why this change is needed. This forces a copy at the interface. I think it's OK to have the caller of this observer decide whether to make a copy of it.

Comment (adisuissa) [reply envoy/config/subscription.h:67]: +1
Why should this be any different than `resource()`, for example?
The caller should ensure that the lifetime of the object is maintained properly.

**Review by adisuissa — COMMENTED** [review]
Thanks!
I'm not sure I fully understand the motivation for this PR. Can you please update the PR description with the relevant background?

/wait

File: envoy/config/subscription.h (hunk 2 above)
Comment (botengyao) [reply envoy/config/subscription.h]: Right, done.

**Comment by botengyao** [comment]
> Thanks! I'm not sure I fully understand the motivation for this PR. Can you please update the PR description with the relevant background?
> 
> /wait

Thanks! updated.

File: source/common/config/decoded_resource_impl.h [hunk 3]
```diff
@@ -77,15 +77,15 @@ class DecodedResourceImpl : public DecodedResource {
...
+    return metadata_.has_value() ? makeOptRef(metadata_.value()) : absl::nullopt;
   }
 
 private:
   DecodedResourceImpl(OpaqueResourceDecoder& resource_decoder, absl::optional<std::string> name,
                       const Protobuf::RepeatedPtrField<std::string>& aliases,
                       const ProtobufWkt::Any& resource, bool has_resource,
                       const std::string& version, absl::optional<std::chrono::milliseconds> ttl,
-                      const OptRef<const envoy::config::core::v3::Metadata> metadata)
+                      const absl::optional<envoy::config::core::v3::Metadata> metadata)
```
Comment (jmarantz) [inline source/common/config/decoded_resource_impl.h:88]: nit: I'm never sure if this style of passing a non-pod by const value will result in an extra copy.

Consider passing by rvalue-ref or const ref.

**Review by jmarantz — APPROVED** [review]

**Review by yanavlasov — APPROVED** [review]
